### PR TITLE
Revamp off-chain worker

### PR DESCRIPTION
**Pull request by satyamakgec**
TODO:
- [x] MESH-1126

- [x] MESH-1128

### CODE REVIEW

File: pallets/cdd-offchain-worker/src/lib.rs [hunk 1]
```diff
@@ -216,49 +253,40 @@ impl<T: Trait> frame_support::unsigned::ValidateUnsigned for Module<T> {
...
+
+            ValidTransaction::with_tag_prefix("CddOffchainWorker")
+                .priority(T::UnsignedPriority::get().saturating_add(
+                    1000 * u64::try_from(payload.nominators.len()).unwrap_or_default(),
+                ))
+                // The transaction is only valid for next 5 blocks. After that it's
+                // going to be revalidated by the pool.
+                .longevity(5)
+                // Not to be propagated to peer nodes.
+                .propagate(false)
```
Comment (maxsam4) [inline pallets/cdd-offchain-worker/src/lib.rs:288]: Why is propagation not allowed?

File: pallets/cdd-offchain-worker/src/lib.rs [hunk 2]
```diff
@@ -206,6 +242,7 @@ impl<T: Trait> Module<T> {
     }
 }
 
+#[allow(deprecated)]
```
Comment (maxsam4) [inline pallets/cdd-offchain-worker/src/lib.rs:245]: Why was this added?

File: pallets/staking/src/lib.rs [hunk 3]
```diff
@@ -2607,35 +2611,26 @@ decl_module! {
...
-        invalid_nominators
+            .map(|(_id, stash)| stash)
+            .collect::<Vec<T::AccountId>>()
     }
 
     /// POLYMESH-NOTE: This is Polymesh specific change.
     /// Here we are assuming that passed targets are always be a those nominators whose cdd
     /// claim get expired or going to expire after the `buffer_time`.
-    pub fn unsafe_validate_cdd_expiry_nominators(targets: Vec<T::AccountId>) -> DispatchResult {
+    fn unchecked_remove_expired_cdd_nominators(targets: &Vec<T::AccountId>) -> DispatchResult {
```
Comment (maxsam4) [inline pallets/staking/src/lib.rs:2633]: This function should make sure that the cdd has actually expired. We shouldn't be trusting a single validator unilaterally. Also, offchain tx can be submitted by anyone, I think.

File: pallets/cdd-offchain-worker/src/lib.rs [hunk 4]
```diff
@@ -216,49 +253,40 @@ impl<T: Trait> frame_support::unsigned::ValidateUnsigned for Module<T> {
...
-                // We can still have multiple transactions compete for the same "spot",
-                // and the one with higher priority will replace other one in the pool.
-                provides: vec![codec::Encode::encode(&(KEY_TYPE.0, last_unsigned_at))],
-                // The transaction is only valid for next n blocks. After that it's
-                // going to be revalidated by the pool. n = cooling interval
-                longevity: (T::CoolingInterval::get()).saturated_into::<u64>(),
-                propagate: true,
-            })
+            ensure!(
+                SignedPayload::<T>::verify::<T::AuthorityId>(payload, signature.clone()),
```
Comment (maxsam4) [inline pallets/cdd-offchain-worker/src/lib.rs:276]: I think this will only check the sign against local keys. If we want the tx to remain local, we don't need to sign it. We can simply check if the tx originated locally using the same approach that phragmen uses. https://github.com/PolymathNetwork/Polymesh/blob/7837422c498f682dfed70eb989ab8abe36a57ca7/pallets/staking/src/lib.rs#L3997

**Review by maxsam4 — CHANGES_REQUESTED** [review]

File: pallets/cdd-offchain-worker/src/lib.rs (hunk 1 above)
Comment (satyamakgec) [reply pallets/cdd-offchain-worker/src/lib.rs:288]: I don't want other nodes to create the transactions, I think other than authored node.

File: pallets/cdd-offchain-worker/src/lib.rs (hunk 2 above)
Comment (satyamakgec) [reply pallets/cdd-offchain-worker/src/lib.rs:245]: I added the declaration from the staking pallet so it comes from there. Initially, I thought something from the rust side that is deprecated in the nightly substrate is using that internally but I just checked it compiles without it as well so we can safely remove it I suppose.

File: pallets/staking/src/lib.rs (hunk 3 above)
Comment (satyamakgec) [reply pallets/staking/src/lib.rs:2633]: FYI - offchain txn only be submitted by the validator only. I did not add the recheck to save the cost as the validator already computed it off-chain

Comment (maxsam4) [reply pallets/staking/src/lib.rs:2633]: I don't think that's true. I think any node can run an offchain worker and submit offchain tx.

File: pallets/cdd-offchain-worker/src/lib.rs (hunk 1 above)
Comment (maxsam4) [reply pallets/cdd-offchain-worker/src/lib.rs:288]: That's not how this will work. All nodes will create the tx regardless of propagation. However, this will be a wasted effort since they won't be able to broadcast their work if they are not author.

File: pallets/staking/src/lib.rs (hunk 3 above)
Comment (satyamakgec) [reply pallets/staking/src/lib.rs:2633]: Yes, you are right any node can run off-chain tx but in our code, we only allowed to run by the validator as we have a check `is_validator()` that allows only validators to compute the data.

Comment (maxsam4) [reply pallets/staking/src/lib.rs:2633]: Except that's in the tx creation phase, not propagation/import phase. therefore, anyone can artificially create the tx and then propograte. The check will not be called again.

File: pallets/cdd-offchain-worker/src/lib.rs (hunk 1 above)
Comment (adamdossa) [reply pallets/cdd-offchain-worker/src/lib.rs:288]: I think by default only authoring nodes will run off-chain workers:
```
        --offchain-worker <ENABLED>
            Should execute offchain workers on every block.
            
            By default it's only enabled for nodes that are authoring new blocks. [default: WhenValidating]  [possible
            values: Always, Never, WhenValidating]
```

File: pallets/staking/src/lib.rs (hunk 3 above)
Comment (adamdossa) [reply pallets/staking/src/lib.rs:2633]: I think, in any case, we should double-check that each proposed nominator really does have an expired CDD check (rather than trusting the submitter, even if it can only be submitted by a validator).

Otherwise, a malicious validator could drop the CDD checks for other validators etc. and cause havoc.

File: pallets/cdd-offchain-worker/src/lib.rs (hunk 1 above)
Comment (maxsam4) [reply pallets/cdd-offchain-worker/src/lib.rs:288]: Yes, from "all nodes", I meant all nodes who are validators. The problem here is that only the node that is authoring the current node will be able to include the transaction if it is not being propagated. Other validator nodes will just be wasting resources.

Comment (adamdossa) [reply pallets/cdd-offchain-worker/src/lib.rs:288]: Is there an easy way to only have the block authoring node run an off-chain worker? In practice these off-chain workers can run every era - so the amount of work should be small.

Comment (maxsam4) [reply pallets/cdd-offchain-worker/src/lib.rs:288]: > Is there an easy way to only have the block authoring node run an off-chain worker? 

Offchain worker and block authoring are independent, asynchronous processes. We can not tie them together. An offchain worker may take multiple blocks of time to execute.
